File: crdt/compact.go
package crdt

// Horizon-based tombstone compaction — the GC pass for long-lived CRDT
// states. The horizon is a stability floor the CALLER guarantees: every
// replica has observed all operations with HLC < horizon, and no in-flight
// operation references addresses older than it. Under that contract:
//
//   - RGA lists drop tombstoned LEAF nodes (no child anchors), cascading to
//     a fixpoint — survivor ordering is untouched by construction.
//   - OR-Sets drop removed tags (and their removal markers) older than the
//     horizon; empty entries are pruned. Add-wins is preserved: only tags
//     already observed-removed are dropped.
//   - Text states never drop addresses — tombstoned fragments are
//     SKELETONIZED (content freed, adjacent skeletons coalesced), keeping
//     every anchor resolvable forever, Yjs GC-marker style.
//
// All Compact methods return how many units they dropped (nodes, tags,
// fragments coalesced away) so callers can meter GC effectiveness. A zero
// horizon is a no-op.

// Compact drops tombstoned leaf nodes older than the horizon, cascading
// until no more can be dropped. Returns the number of nodes removed.
func (l *RGAListState) Compact(before HLC) int {
	if before.IsZero() || len(l.Nodes) == 0 {
		return 0
	}
	dropped := 0
	for {
		// Recompute anchors each round: dropping a leaf may expose its parent.
		hasChild := make(map[string]bool, len(l.Nodes))
		for _, node := range l.Nodes {
			hasChild[rgaNodeKey(node.ParentID)] = true
		}
		droppedThisRound := 0
		for key, node := range l.Nodes {
			if node.Tombstone && !hasChild[key] && before.After(node.ID) {
				delete(l.Nodes, key)
				droppedThisRound++
			}
		}
		if droppedThisRound == 0 {
			break
		}
		dropped += droppedThisRound
	}
	if dropped > 0 {
		l.order = nil
	}
	return dropped
}

// Compact drops observed-removed tags older than the horizon along with
// their removal markers, pruning entries left tagless. Returns the number
// of tags dropped.
func (s *ORSetState) Compact(before HLC) int {
	if before.IsZero() {
		return 0
	}
	dropped := 0
	for elem, tags := range s.Entries {
		kept := tags[:0]
		for _, tag := range tags {
			if s.tagRemoved(elem, tag) && before.After(tag.HLC) {
				delete(s.Removed, removedKey(elem, tag))
				delete(s.Removed, tagKey(tag))
				dropped++
				continue
			}
			kept = append(kept, tag)
		}
		if len(kept) == 0 {
			delete(s.Entries, elem)
		} else {
			s.Entries[elem] = kept
		}
	}
	return dropped
}

// Compact skeletonizes tombstoned fragments whose origin is older than the
// horizon: content is freed and adjacent skeletons coalesce, but addresses
// are preserved so every anchor stays resolvable. Returns the number of
// fragments freed or coalesced away.
func (t *TextState) Compact(before HLC) int {
	if before.IsZero() {
		return 0
	}
	dropped := 0
	for key, frags := range t.Frags {
		changed := false
		for _, f := range frags {
			if f.Tombstone && f.Content != "" && before.After(f.Origin) {
				f.Content = ""
				f.Attrs = nil
				changed = true
				dropped++
			}
		}
		if !changed && len(frags) < 2 {
			continue
		}
		// Coalesce adjacent skeletons (contiguous, both tombstoned+empty).
		out := frags[:0]
		for _, f := range frags {
			if n := len(out); n > 0 {
				prev := out[n-1]
				if prev.Tombstone && prev.Content == "" && f.Tombstone && f.Content == "" &&
					prev.Start+prev.Length == f.Start {
					prev.Length += f.Length
					dropped++
					continue
				}
			}
			out = append(out, f)
		}
		t.Frags[key] = out
	}
	if dropped > 0 {
		t.segs = nil
	}
	return dropped
}

// Compact walks every field of a record state, compacting the types that
// accumulate tombstones. Returns the total units dropped.
func (st *State) Compact(before HLC) int {
	if st == nil || before.IsZero() {
		return 0
	}
	dropped := 0
	for _, fs := range st.Fields {
		switch fs.Type {
		case TypeList:
			if fs.ListState != nil {
				dropped += fs.ListState.Compact(before)
			}
		case TypeSet:
			if fs.SetState != nil {
				dropped += fs.SetState.Compact(before)
			}
		case TypeText:
			if fs.TextState != nil {
				dropped += fs.TextState.Compact(before)
			}
		}
	}
	return dropped
}

File: crdt-js/src/compact.ts
/**
 * Horizon-based tombstone compaction. Port of crdt/compact.go, written
 * pure so it composes with copy-on-write state.
 *
 * The horizon is a stability floor the CALLER guarantees: every replica
 * has observed all operations with HLC < horizon, and no in-flight
 * operation references addresses older than it. Passing a horizon that
 * does not hold will diverge replicas.
 */

import type {
  RGAListState, ORSetState, ORSetTag, TextState, TextFragment,
  DocumentState, FieldState, HLC,
} from "./types.js";
import { hlcAfter, hlcIsZero, hlcString } from "./hlc.js";
import { removedKey, tagKey } from "./merge.js";

/**
 * Drop tombstoned LEAF nodes older than the horizon, cascading until no
 * more can be dropped. Survivor ordering is untouched by construction: a
 * node with children is never removed, so no anchor is ever orphaned.
 */
export function compactListState(
  state: RGAListState,
  before: HLC
): { state: RGAListState; dropped: number } {
  if (hlcIsZero(before) || Object.keys(state.nodes).length === 0) {
    return { state, dropped: 0 };
  }

  let nodes = state.nodes;
  let dropped = 0;

  for (;;) {
    // Recompute anchors each round: dropping a leaf may expose its parent.
    const hasChild = new Set<string>();
    for (const node of Object.values(nodes)) {
      hasChild.add(hlcString(node.parent_id));
    }

    const survivors: Record<string, typeof nodes[string]> = {};
    let droppedThisRound = 0;
    for (const [key, node] of Object.entries(nodes)) {
      if (node.tombstone && !hasChild.has(key) && hlcAfter(before, node.id)) {
        droppedThisRound++;
        continue;
      }
      survivors[key] = node;
    }

    if (droppedThisRound === 0) break;
    nodes = survivors;
    dropped += droppedThisRound;
  }

  return dropped === 0 ? { state, dropped: 0 } : { state: { nodes }, dropped };
}

/** True when this tag has been observed-removed for this element. */
function tagRemoved(
  removed: Record<string, boolean>,
  elem: string,
  tag: ORSetTag
): boolean {
  return Boolean(removed[removedKey(elem, tag)] || removed[tagKey(tag)]);
}

/**
 * Drop observed-removed tags older than the horizon along with their
 * element-scoped removal markers, pruning entries left tagless.
 *
 * Only the element-scoped marker is consumed. A legacy tag-only marker is
 * SHARED evidence — a multi-element add shares one tag — so deleting it
 * would resurrect sibling elements that still rely on it, and which
 * sibling would depend on iteration order. The tiny legacy marker stays.
 */
export function compactSetState(
  state: ORSetState,
  before: HLC
): { state: ORSetState; dropped: number } {
  if (hlcIsZero(before)) return { state, dropped: 0 };

  const entries: Record<string, ORSetTag[]> = {};
  const removed = { ...state.removed };
  let dropped = 0;

  for (const [elem, tags] of Object.entries(state.entries)) {
    const kept: ORSetTag[] = [];
    for (const tag of tags) {
      if (tagRemoved(state.removed, elem, tag) && hlcAfter(before, tag.hlc)) {
        delete removed[removedKey(elem, tag)];
        dropped++;
        continue;
      }
      kept.push(tag);
    }
    if (kept.length > 0) entries[elem] = kept;
  }

  return dropped === 0 ? { state, dropped: 0 } : { state: { entries, removed }, dropped };
}

/**
 * Skeletonize tombstoned fragments whose origin is older than the horizon:
 * content is freed and adjacent skeletons coalesce, but addresses are
 * PRESERVED so every cursor anchor stays resolvable.
 */
export function compactTextState(
  state: TextState,
  before: HLC
): { state: TextState; dropped: number } {
  if (hlcIsZero(before)) return { state, dropped: 0 };

  const frags: Record<string, TextFragment[]> = {};
  let dropped = 0;

  for (const [key, list] of Object.entries(state.frags)) {
    const skeletonized = list.map((f) => {
      if (f.tombstone && f.content !== "" && hlcAfter(before, f.origin)) {
        dropped++;
        const { attrs: _attrs, ...rest } = f;
        return { ...rest, content: "" };
      }
      return f;
    });

    // Coalesce adjacent skeletons (contiguous, both tombstoned and empty).
    const out: TextFragment[] = [];
    for (const f of skeletonized) {
      const prev = out[out.length - 1];
      if (
        prev && prev.tombstone && prev.content === "" &&
        f.tombstone && f.content === "" &&
        prev.start + prev.length === f.start
      ) {
        out[out.length - 1] = { ...prev, length: prev.length + f.length };
        dropped++;
        continue;
      }
      out.push(f);
    }
    frags[key] = out;
  }

  return dropped === 0 ? { state, dropped: 0 } : { state: { frags }, dropped };
}

/** Compact every compactable field of one document. */
export function compactDocument(
  doc: DocumentState,
  before: HLC
): { doc: DocumentState; dropped: number } {
  if (hlcIsZero(before)) return { doc, dropped: 0 };

  const fields: Record<string, FieldState> = {};
  let dropped = 0;

  for (const [name, fs] of Object.entries(doc.fields)) {
    if (fs.type === "list" && fs.list_state) {
      const r = compactListState(fs.list_state, before);
      dropped += r.dropped;
      fields[name] = r.dropped ? { ...fs, list_state: r.state } : fs;
    } else if (fs.type === "set" && fs.set_state) {
      const r = compactSetState(fs.set_state, before);
      dropped += r.dropped;
      fields[name] = r.dropped ? { ...fs, set_state: r.state } : fs;
    } else if (fs.type === "text" && fs.text_state) {
      const r = compactTextState(fs.text_state, before);
      dropped += r.dropped;
      fields[name] = r.dropped ? { ...fs, text_state: r.state } : fs;
    } else {
      fields[name] = fs;
    }
  }

  return dropped === 0 ? { doc, dropped: 0 } : { doc: { ...doc, fields }, dropped };
}
